src/optimizer/build_probe_side_optimizer.cpp

```diff
@@ -101,6 +85,42 @@ BuildSize BuildProbeSideOptimizer::GetBuildSizes(LogicalOperator &op) {
 	return ret;
 }
 
+double BuildProbeSideOptimizer::GetBuildSize(vector<LogicalType> types, const idx_t cardinality) {
+	// Row width in the hash table
+	types.push_back(LogicalType::HASH);
+	auto tuple_layout = TupleDataLayout();
+	tuple_layout.Initialize(types);
+	auto row_width = tuple_layout.GetRowWidth();
+
+	for (const auto &type : types) {
+		TypeVisitor::VisitReplace(type, [&](const LogicalType &visited_type) {
+			// Penalty for variable-size types (we don't have statistics here yet)
+			switch (visited_type.InternalType()) {
+			case PhysicalType::VARCHAR:
+				row_width += 8;
+				break;
+			case PhysicalType::LIST:
```

Comment:
I know this has already been merged, but maybe `PhysicalType::Struct` can be added?

Reply:
For structs we should add the bits that are used for the validity mask, but the recursion of `TypeVisitor` handles the child vectors anyway so this should be pretty accurate already.